### PR TITLE
Fix some crashes and using null pointers

### LINKED ISSUE #30343: Project crash in AreaBullet::get_spOv_mode
**Godot version:**
3.2.dev.custom_build. 550f436f8

**OS/device including version:**
Ubuntu 19.04

**Issue description:**
Backtrace
```
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x43f60) [0x7f2aa598af60] (??:0)
[2] AreaBullet::get_spOv_mode() (/home/rafal/Pulpit/godot/modules/bullet/area_bullet.h:119)
[3] RigidBodyBullet::reload_space_override_modificator() (/home/rafal/Pulpit/godot/modules/bullet/rigid_body_bullet.cpp:921)
[4] RigidBodyBullet::dispatch_callbacks() (/home/rafal/Pulpit/godot/modules/bullet/rigid_body_bullet.cpp:385)
[5] SpaceBullet::flush_queries() (/home/rafal/Pulpit/godot/modules/bullet/space_bullet.cpp:358 (discriminator 2))
[6] onBulletPreTickCallback(btDynamicsWorld*, float) (/home/rafal/Pulpit/godot/modules/bullet/space_bullet.cpp:539)
[7] btDiscreteDynamicsWorld::internalSingleStepSimulation(float) (/home/rafal/Pulpit/godot/thirdparty/bullet/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:462)
[8] btSoftRigidDynamicsWorld::internalSingleStepSimulation(float) (/home/rafal/Pulpit/godot/thirdparty/bullet/BulletSoftBody/btSoftRigidDynamicsWorld.cpp:91)
[9] btDiscreteDynamicsWorld::stepSimulation(float, int, float) (/home/rafal/Pulpit/godot/thirdparty/bullet/BulletDynamics/Dynamics/btDiscreteDynamicsWorld.cpp:435 (discriminator 2))
[10] SpaceBullet::step(float) (/home/rafal/Pulpit/godot/modules/bullet/space_bullet.cpp:366)
[11] BulletPhysicsServer::step(float) (/home/rafal/Pulpit/godot/modules/bullet/bullet_physics_server.cpp:1566 (discriminator 2))
[12] Main::iteration() (/home/rafal/Pulpit/godot/main/main.cpp:1909)
[13] OS_X11::run() (/home/rafal/Pulpit/godot/platform/x11/os_x11.cpp:3159)
[14] /usr/bin/godot(main+0xdf) [0x139e581] (/home/rafal/Pulpit/godot/platform/x11/godot_x11.cpp:56)
[15] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7f2aa596db6b] (??:0)
[16] /usr/bin/godot(_start+0x2a) [0x139e3ea] (??:?)
```

**Steps to reproduce:**
1. Read informations about project - https://github.com/qarmin/The-worst-Godot-test-project#informations
2. Run minimal project and wait(happens very rarely)

**Minimal reproduction project:**
Project is very big, because I don't know what exactly cause crash, and it happens rarely
https://github.com/qarmin/The-worst-Godot-test-project/commit/6ef8a28ebb1533c4017806c5211d104c47c875da
ZIP - https://github.com/qarmin/The-worst-Godot-test-project/archive/6ef8a28ebb1533c4017806c5211d104c47c875da.zip

### LINKED ISSUE #32658: Godot crash when using PackedDataContainer::size()
**Godot version:**
3.2.alpha.custom_build. 02d75f99b

**OS/device including version:**
Ubuntu 19.04
**Issue description:**
Backtrace
```
core/pool_vector.h:307:75: runtime error: reference binding to null pointer of type 'const unsigned char'
core/io/marshalls.h:149:16: runtime error: load of null pointer of type 'const uint8_t'
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] godots() [0x14052b0] (/home/rafal/Pulpit/mojgodot/platform/x11/crash_handler_x11.cpp:54)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43f60) [0x7f47a3686f60] (??:0)
[3] godots() [0xf536ef5] (/home/rafal/Pulpit/mojgodot/./core/io/marshalls.h:149)
[4] PackedDataContainer::_size(unsigned int) const (/home/rafal/Pulpit/mojgodot/core/packed_data_container.cpp:141)
[5] PackedDataContainer::size() const (/home/rafal/Pulpit/mojgodot/core/packed_data_container.cpp:48)
[6] MethodBind0RC<int>::call(Object*, Variant const**, int, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/./core/method_bind.gen.inc:593 (discriminator 4))
[7] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/core/object.cpp:921 (discriminator 1))
[8] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/core/variant_call.cpp:1083 (discriminator 1))
[9] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) (/home/rafal/Pulpit/mojgodot/modules/gdscript/gdscript_function.cpp:1079)
[10] GDScriptInstance::call_multilevel(StringName const&, Variant const**, int) (/home/rafal/Pulpit/mojgodot/modules/gdscript/gdscript.cpp:1180)
[11] Node::_notification(int) (/home/rafal/Pulpit/mojgodot/scene/main/node.cpp:58)
[12] Node::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/main/node.h:46 (discriminator 14))
[13] CanvasItem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/2d/canvas_item.h:166 (discriminator 3))
[14] Node2D::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/2d/node_2d.h:38 (discriminator 3))
[15] Object::notification(int, bool) (/home/rafal/Pulpit/mojgodot/core/object.cpp:933)
[16] SceneTree::_notify_group_pause(StringName const&, int) (/home/rafal/Pulpit/mojgodot/scene/main/scene_tree.cpp:964)
[17] SceneTree::idle(float) (/home/rafal/Pulpit/mojgodot/scene/main/scene_tree.cpp:516 (discriminator 3))
[18] Main::iteration() (/home/rafal/Pulpit/mojgodot/main/main.cpp:1978)
[19] OS_X11::run() (/home/rafal/Pulpit/mojgodot/platform/x11/os_x11.cpp:3192)
[20] godots(main+0x342) [0x13fcba4] (/home/rafal/Pulpit/mojgodot/platform/x11/godot_x11.cpp:57)
[21] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7f47a3669b6b] (??:0)
[22] godots(_start+0x2a) [0x13fc7aa] (??:?)
```
**Steps to reproduce:**
Read project manual - https://github.com/qarmin/The-worst-Godot-test-project#the-worst-godot-test-project
Run project
**Minimal reproduction project:**
[The-worst-Godot-test-project.zip](https://github.com/godotengine/godot/files/3704229/The-worst-Godot-test-project.zip)

### LINKED ISSUE #32659: Godot crash when using Image::expand_x2_hq2x
**Godot version:**
3.2.alpha.custom_build. 02d75f99b

**OS/device including version:**
Ubuntu 19.04
**Issue description:**
Backtrace
```
thirdparty/misc/hq2x.cpp:261:11: runtime error: load of null pointer of type 'const uint32_t'
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] godots() [0x14052b0] (/home/rafal/Pulpit/mojgodot/platform/x11/crash_handler_x11.cpp:54)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43f60) [0x7f60ff232f60] (??:0)
[3] hq2x_resize(unsigned int const*, unsigned int, unsigned int, unsigned int*, unsigned int, unsigned int, unsigned int, unsigned int, bool, bool) (/home/rafal/Pulpit/mojgodot/thirdparty/misc/hq2x.cpp:261)
[4] Image::expand_x2_hq2x() (/home/rafal/Pulpit/mojgodot/core/image.cpp:1325)
[5] MethodBind0::call(Object*, Variant const**, int, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/./core/method_bind.gen.inc:61 (discriminator 4))
[6] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/core/object.cpp:921 (discriminator 1))
[7] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/core/variant_call.cpp:1083 (discriminator 1))
[8] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) (/home/rafal/Pulpit/mojgodot/modules/gdscript/gdscript_function.cpp:1085)
[9] GDScriptInstance::call_multilevel(StringName const&, Variant const**, int) (/home/rafal/Pulpit/mojgodot/modules/gdscript/gdscript.cpp:1180)
[10] Node::_notification(int) (/home/rafal/Pulpit/mojgodot/scene/main/node.cpp:58)
[11] Node::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/main/node.h:46 (discriminator 14))
[12] CanvasItem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/2d/canvas_item.h:166 (discriminator 3))
[13] Node2D::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/2d/node_2d.h:38 (discriminator 3))
[14] Object::notification(int, bool) (/home/rafal/Pulpit/mojgodot/core/object.cpp:933)
[15] SceneTree::_notify_group_pause(StringName const&, int) (/home/rafal/Pulpit/mojgodot/scene/main/scene_tree.cpp:964)
[16] SceneTree::idle(float) (/home/rafal/Pulpit/mojgodot/scene/main/scene_tree.cpp:516 (discriminator 3))
[17] Main::iteration() (/home/rafal/Pulpit/mojgodot/main/main.cpp:1978)
[18] OS_X11::run() (/home/rafal/Pulpit/mojgodot/platform/x11/os_x11.cpp:3192)
[19] godots(main+0x342) [0x13fcba4] (/home/rafal/Pulpit/mojgodot/platform/x11/godot_x11.cpp:57)
[20] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7f60ff215b6b] (??:0)
[21] godots(_start+0x2a) [0x13fc7aa] (??:?)
```
**Steps to reproduce:**
Run project
**Minimal reproduction project:**
[The-worst-Godot-test-project.zip](https://github.com/godotengine/godot/files/3704317/The-worst-Godot-test-project.zip)

### LINKED ISSUE #32754: Godot crash when using TranslationServer::get_loaded_locales
**Godot version:**
3.2.alpha.custom_build. 3cc94b2c0
**OS/device including version:**
Ubuntu 19.04
**Issue description:**
Backtrace
```
core/translation.h:55:52: runtime error: member access within null pointer of type 'const struct Translation'
core/translation.h:55:52: runtime error: member access within null pointer of type 'const struct Translation'
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] godots() [0x140a2b0] (/home/rafal/Pulpit/mojgodot/platform/x11/crash_handler_x11.cpp:54)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43f60) [0x7f0435277f60] (??:0)
[3] CowData<wchar_t>::_ref(CowData<wchar_t> const&) (/home/rafal/Pulpit/mojgodot/./core/cowdata.h:347)
[4] String::String(String const&) (/home/rafal/Pulpit/mojgodot/./core/ustring.h:355)
[5] Translation::get_locale() const (/home/rafal/Pulpit/mojgodot/./core/translation.h:55)
[6] TranslationServer::get_loaded_locales() const (/home/rafal/Pulpit/mojgodot/core/translation.cpp:991)
[7] MethodBind0RC<Array>::call(Object*, Variant const**, int, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/./core/method_bind.gen.inc:593 (discriminator 4))
[8] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/core/object.cpp:921 (discriminator 1))
[9] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/home/rafal/Pulpit/mojgodot/core/variant_call.cpp:1096 (discriminator 1))
[10] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) (/home/rafal/Pulpit/mojgodot/modules/gdscript/gdscript_function.cpp:1079)
[11] GDScriptInstance::call_multilevel(StringName const&, Variant const**, int) (/home/rafal/Pulpit/mojgodot/modules/gdscript/gdscript.cpp:1180)
[12] Node::_notification(int) (/home/rafal/Pulpit/mojgodot/scene/main/node.cpp:58)
[13] Node::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/main/node.h:46 (discriminator 14))
[14] CanvasItem::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/2d/canvas_item.h:166 (discriminator 3))
[15] Node2D::_notificationv(int, bool) (/home/rafal/Pulpit/mojgodot/./scene/2d/node_2d.h:38 (discriminator 3))
[16] Object::notification(int, bool) (/home/rafal/Pulpit/mojgodot/core/object.cpp:933)
[17] SceneTree::_notify_group_pause(StringName const&, int) (/home/rafal/Pulpit/mojgodot/scene/main/scene_tree.cpp:964)
[18] SceneTree::idle(float) (/home/rafal/Pulpit/mojgodot/scene/main/scene_tree.cpp:516 (discriminator 3))
[19] Main::iteration() (/home/rafal/Pulpit/mojgodot/main/main.cpp:1978)
[20] OS_X11::run() (/home/rafal/Pulpit/mojgodot/platform/x11/os_x11.cpp:3192)
[21] godots(main+0x342) [0x1401ba4] (/home/rafal/Pulpit/mojgodot/platform/x11/godot_x11.cpp:57)
[22] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7f043525ab6b] (??:0)
[23] godots(_start+0x2a) [0x14017aa] (??:?)

```
**Steps to reproduce:**
Read project manual - https://github.com/qarmin/The-worst-Godot-test-project#the-worst-godot-test-project
Run project
**Minimal reproduction project:**
[The-worst-Godot-test-project.zip](https://github.com/godotengine/godot/files/3719138/The-worst-Godot-test-project.zip)

### LINKED ISSUE #32877: Godot read memory outside of array in RasterizerSceneGLES3::_render_geometry
**Godot version:**
3.2.alpha.custom_build. a1033aea5
**OS/device including version:**
Ubuntu 19.04
**Issue description:**
To fix this should be enough to check enum value of PrimitiveType probably in ImmediateGeometry
::begin
```
core/command_queue_mt.h:301:2: runtime error: load of value 4294926790, which is not a valid value for type 'PrimitiveType'
drivers/gles3/rasterizer_storage_gles3.h:839:10: runtime error: load of value 4294926790, which is not a valid value for type 'PrimitiveType'
drivers/gles3/rasterizer_scene_gles3.cpp:1740:33: runtime error: load of value 4294965270, which is not a valid value for type 'PrimitiveType'
drivers/gles3/rasterizer_scene_gles3.cpp:1740:42: runtime error: index -2026 out of bounds for type 'unsigned int [7]'
drivers/gles3/rasterizer_scene_gles3.cpp:1740:33: runtime error: load of value 4294965270, which is not a valid value for type 'PrimitiveType'
=================================================================
==21872==ERROR: AddressSanitizer: global-buffer-overflow on address 0x000010f27338 at pc 0x00000515330a bp 0x7ff7b5e94910 sp 0x7ff7b5e94900
READ of size 4 at 0x000010f27338 thread T3
    #0 0x5153309 in RasterizerSceneGLES3::_render_geometry(RasterizerSceneGLES3::RenderList::Element*) (/usr/bin/godots+0x5153309)
    #1 0x5063592 in RasterizerSceneGLES3::_render_list(RasterizerSceneGLES3::RenderList::Element**, int, Transform const&, CameraMatrix const&, unsigned int, bool, bool, bool, bool, bool) drivers/gles3/rasterizer_scene_gles3.cpp:2265
    #2 0x50df8a6 in RasterizerSceneGLES3::render_scene(Transform const&, CameraMatrix const&, bool, RasterizerScene::InstanceBase**, int, RID*, int, RID*, int, RID, RID, RID, RID, int) drivers/gles3/rasterizer_scene_gles3.cpp:4234
    #3 0xd61ad0d in VisualServerScene::_render_scene(Transform, CameraMatrix const&, bool, RID, RID, RID, RID, int) servers/visual/visual_server_scene.cpp:2187
    #4 0xd608a20 in VisualServerScene::render_camera(RID, RID, Vector2, RID) servers/visual/visual_server_scene.cpp:1757
    #5 0xd6f1aea in VisualServerViewport::_draw_3d(VisualServerViewport::Viewport*, ARVRInterface::Eyes) servers/visual/visual_server_viewport.cpp:74
    #6 0xd6f3e6f in VisualServerViewport::_draw_viewport(VisualServerViewport::Viewport*, ARVRInterface::Eyes) servers/visual/visual_server_viewport.cpp:106
    #7 0xd6fff71 in VisualServerViewport::draw_viewports() servers/visual/visual_server_viewport.cpp:343
    #8 0xd59f491 in VisualServerRaster::draw(bool, double) servers/visual/visual_server_raster.cpp:108
    #9 0xd72ec84 in VisualServerWrapMT::thread_draw(bool, double) servers/visual/visual_server_wrap_mt.cpp:44
    #10 0xd9415db in CommandQueueMT::Command2<VisualServerWrapMT, void (VisualServerWrapMT::*)(bool, double), bool, double>::call() core/command_queue_mt.h:301
    #11 0xd3211a4 in CommandQueueMT::flush_one(bool) core/command_queue_mt.h:426
    #12 0xd321ad2 in CommandQueueMT::wait_and_flush_one() core/command_queue_mt.h:459
    #13 0xd72fad9 in VisualServerWrapMT::thread_loop() servers/visual/visual_server_wrap_mt.cpp:72
    #14 0xd72ee38 in VisualServerWrapMT::_thread_callback(void*) servers/visual/visual_server_wrap_mt.cpp:57
    #15 0x4fdfb77 in ThreadPosix::thread_callback(void*) drivers/unix/thread_posix.cpp:74
    #16 0x7ff7dad8c181 in start_thread /build/glibc-KRRWSm/glibc-2.29/nptl/pthread_create.c:486
    #17 0x7ff7d9ffbb1e in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11db1e)

0x000010f27338 is located 8 bytes to the left of global variable '*.LC203' defined in 'drivers/gles3/rasterizer_scene_gles3.cpp' (0x10f27340) of size 20
  '*.LC203' is ascii string 'SsaoBlurShaderGLES3'
0x000010f27338 is located 42 bytes to the right of global variable '*.LC202' defined in 'drivers/gles3/rasterizer_scene_gles3.cpp' (0x10f27300) of size 14
  '*.LC202' is ascii string 'source_normal'
SUMMARY: AddressSanitizer: global-buffer-overflow (/usr/bin/godots+0x5153309) in RasterizerSceneGLES3::_render_geometry(RasterizerSceneGLES3::RenderList::Element*)
```
**Minimal reproduction project:**
[The-worst-Godot-test-project.tar.gz](https://github.com/godotengine/godot/files/3735576/The-worst-Godot-test-project.tar.gz)

### PR DESCRIPTION
**Pull request by qarmin**
Fixes #30343 - I'm not sure about fix, but for now this is better than crash
Fixes #32877
Fixes #32658
Fixes #32754 
Fixes #32659

Enums used in set_refraction_texture_channel and set_ao_texture_channel must be checked to prevent Godot crash.

I changed name of `p_rimitive` parameter to `p_primitive` 

This fixes also this messages from ubsan
core/pool_vector.h:307:75: runtime error: reference binding to null pointer of type 'const unsigned char'
drivers/gles2/rasterizer_storage_gles2.cpp:3245:2: runtime error: null pointer passed as argument 1, which is declared to never be null
drivers/gles2/rasterizer_storage_gles2.cpp:3245:2: runtime error: null pointer passed as argument 2, which is declared to never be null
scene/resources/material.cpp:1731:24: runtime error: index -3 out of bounds for type 'Plane [5]'

### CODE REVIEW

File: drivers/gles2/rasterizer_storage_gles2.cpp [hunk 1]
```diff
@@ -3236,12 +3237,14 @@ Color RasterizerStorageGLES2::multimesh_instance_get_custom_data(RID p_multimesh
 void RasterizerStorageGLES2::multimesh_set_as_bulk_array(RID p_multimesh, const PoolVector<float> &p_array) {
 	MultiMesh *multimesh = multimesh_owner.getornull(p_multimesh);
 	ERR_FAIL_COND(!multimesh);
+	ERR_FAIL_COND(!multimesh->data.ptrw());
```
Comment (akien-mga) [inline drivers/gles2/rasterizer_storage_gles2.cpp]: Didn't you mean checking `!multimesh->data` here?

Comment (qarmin) [reply drivers/gles2/rasterizer_storage_gles2.cpp]: No, data variable is Vector<float> type, and when Vector is empty ptrw() returns NULL.
I could use condition - `ERR_FAIL_COND(multimesh->data.size() == 0);` but I want to do it in a similar way as it is used - copymem function in line 3248.

Comment (akien-mga) [reply drivers/gles2/rasterizer_storage_gles2.cpp]: `ptrw()` does a copy on write though, the "w" stands for "write" as it's the one you should use to write data. If you only want to check that there is data, I think you can use `ptr()` (or indeed `.size() == 0`).

Comment (qarmin) [reply drivers/gles2/rasterizer_storage_gles2.cpp]: Done

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**